File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/ExamListAdapter.java
package de.tum.in.tumcampusapp.component.tumui.grades;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.DateFormat;
import java.util.Collections;
import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.other.generic.adapter.SimpleStickyListHeadersAdapter;
import de.tum.in.tumcampusapp.component.tumui.grades.model.Exam;

/**
 * Custom UI adapter for a list of exams.
 */
public class ExamListAdapter extends SimpleStickyListHeadersAdapter<Exam> {
    private static final DateFormat DF = DateFormat.getDateInstance(DateFormat.MEDIUM);

    ExamListAdapter(Context context, List<Exam> results) {
        super(context, results);
        Collections.sort(infoList);
    }

    @Override
    public String genenrateHeaderName(Exam item) {
        String headerText = super.genenrateHeaderName(item);
        int year = Integer.parseInt(headerText.substring(0, 2));
        if (headerText.charAt(2) == 'W') {
            return context.getString(R.string.winter_semester, year, year + 1);
        } else {
            return context.getString(R.string.summer_semester, year);
        }
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        View view = convertView;

        // find and init UI
        if (view == null) {
            view = mInflater.inflate(R.layout.activity_grades_listview, parent, false);
            holder = new ViewHolder();
            holder.tvName = view.findViewById(R.id.courseNameTextView);
            holder.tvGrade = view.findViewById(R.id.gradeTextView);
            holder.tvDetails1 = view.findViewById(R.id.examDateTextView);
            holder.tvDetails2 = view.findViewById(R.id.examinerTextView);
            holder.gradeBackground = context.getResources()
                                            .getDrawable(R.drawable.grade_background);
            holder.tvGrade.setBackground(holder.gradeBackground);
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }
        // fill UI with data
        Exam exam = infoList.get(position);
        if (exam != null) {
            holder.tvName.setText(exam.getCourse());
            holder.tvGrade.setText(exam.getGrade());
            holder.gradeBackground.setTint(exam.getGradeColor(context));

            holder.tvDetails1.setText(String.format("%s: %s, ", context.getString(R.string.date), DF.format(exam.getDate())));

            holder.tvDetails2.setText(String.format("%s: %s, " +
                                                    "%s: %s",
                                                    context.getString(R.string.examiner), exam.getExaminer(),
                                                    context.getString(R.string.mode), exam.getModus()));
        }

        return view;
    }

    static class ViewHolder {
        TextView tvDetails1;
        TextView tvDetails2;
        TextView tvGrade;
        TextView tvName;
        Drawable gradeBackground;
    }
}

File: app/src/main/java/de/uos/campusapp/component/tumui/grades/ExamListAdapter.kt
package de.uos.campusapp.component.tumui.grades

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import de.uos.campusapp.R
import de.uos.campusapp.component.other.generic.adapter.SimpleStickyListHeadersAdapter
import de.uos.campusapp.component.tumui.grades.model.AbstractExam
import org.joda.time.format.DateTimeFormat

/**
 * Custom UI adapter for a list of exams.
 */
class ExamListAdapter(context: Context, results: List<AbstractExam>) : SimpleStickyListHeadersAdapter<AbstractExam>(context, results.toMutableList()) {

    init {
        itemList.sort()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        val view: View

        if (convertView == null) {
            view = inflater.inflate(R.layout.activity_grades_listview, parent, false)
            holder = ViewHolder(view)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as ViewHolder
        }

        val exam = itemList[position]
        holder.nameTextView.text = exam.course
        holder.gradeTextView.text = exam.gradeString

        val gradeColor = exam.getGradeColor(context)
        holder.gradeTextView.background.setTint(gradeColor)

        val date: String = if (exam.date == null) {
            context.getString(R.string.not_specified)
        } else {
            DATE_FORMAT.print(exam.date)
        }
        holder.examDateTextView.text = String.format("%s: %s", context.getString(R.string.date), date)

        holder.additionalInfoTextView.text = String.format("%s: %s, %s: %s",
                context.getString(R.string.examiner),
                exam.examiner ?: context.getString(R.string.not_specified),
                context.getString(R.string.type),
                exam.type ?: context.getString(R.string.not_specified)
        )

        return view
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    internal class ViewHolder(itemView: View) {
        var nameTextView: TextView = itemView.findViewById(R.id.courseNameTextView)
        var gradeTextView: TextView = itemView.findViewById(R.id.gradeTextView)
        var examDateTextView: TextView = itemView.findViewById(R.id.examDateTextView)
        var additionalInfoTextView: TextView = itemView.findViewById(R.id.additionalInfoTextView)
    }

    companion object {
        private val DATE_FORMAT = DateTimeFormat.mediumDate()
    }
}
